File: backend-ruby/app/controllers/api/v1/dang_ky_thi_controller.rb
class Api::V1::DangKyThiController < ApplicationController
  before_action :set_dang_ky_thi, only: [:show, :update, :destroy, :cancel]

  def index
    records = DangKyThi
              .includes(:ky_thi, :sinh_vien)
              .order(DangKyThiID: :asc)

    render json: records.as_json(
      include: {
        ky_thi: {
          only: [:KyThiID, :MaKyThi, :TenKyThi, :TrangThai]
        },
        sinh_vien: {
          only: [:SinhVienID, :MaSinhVien, :HoTen, :Lop]
        }
      }
    )
  end

  def show
    render json: @dang_ky_thi.as_json(
      include: {
        ky_thi: {
          only: [:KyThiID, :MaKyThi, :TenKyThi, :TrangThai]
        },
        sinh_vien: {
          only: [:SinhVienID, :MaSinhVien, :HoTen, :Lop]
        }
      }
    )
  end

  def create
    ky_thi = KyThi.find(dang_ky_thi_params[:KyThiID])

    unless ky_thi.TrangThai == "published"
      return render json: {
        error: "Only published exams can accept registrations."
      }, status: :unprocessable_entity
    end

    record = DangKyThi.new(dang_ky_thi_params)
    record.TrangThaiDangKy ||= "registered"
    record.SoBaoDanh ||= generate_so_bao_danh(record.KyThiID)

    if record.save
      render json: record, status: :created
    else
      render json: { errors: record.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotFound
    render json: { error: "Exam or student not found." }, status: :not_found
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "This student is already registered for this exam or the exam number already exists." }, status: :conflict
  end

  def update
    if @dang_ky_thi.update(dang_ky_thi_params)
      render json: @dang_ky_thi
    else
      render json: { errors: @dang_ky_thi.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "Duplicate registration or exam number." }, status: :conflict
  end

  def destroy
    @dang_ky_thi.destroy
    render json: { message: "Deleted successfully." }
  rescue ActiveRecord::InvalidForeignKey
    render json: { error: "Cannot delete because this registration is already used in room assignment, seating, or attendance." }, status: :conflict
  end

  def cancel
    if @dang_ky_thi.ky_thi.TrangThai != "published"
      return render json: {
        error: "Registration can only be cancelled before room assignment."
      }, status: :unprocessable_entity
    end

    @dang_ky_thi.update!(TrangThaiDangKy: "cancelled")
    render json: @dang_ky_thi
  end

  private

  def set_dang_ky_thi
    @dang_ky_thi = DangKyThi.find(params[:id])
  end

  def dang_ky_thi_params
    params.require(:dang_ky_thi).permit(
      :KyThiID,
      :SinhVienID,
      :SoBaoDanh,
      :TrangThaiDangKy
    )
  end

  def generate_so_bao_danh(ky_thi_id)
    next_number = DangKyThi.where(KyThiID: ky_thi_id).count + 1
    "SBD#{ky_thi_id.to_s.rjust(3, '0')}#{next_number.to_s.rjust(3, '0')}"
  end
end

File: backend-csharp/Controllers/DangKyThiController.cs
using System.Text.Json;
using backend_csharp.Data;
using backend_csharp.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/dang_ky_thi")]
public class DangKyThiController(ExamDbContext dbContext) : ApiControllerBase
{
    [HttpGet]
    public async Task<IActionResult> Index()
    {
        List<DangKyThi> records = await dbContext.DangKyThi
            .Include(record => record.KyThi)
            .Include(record => record.SinhVien)
            .OrderBy(record => record.DangKyThiID)
            .ToListAsync();

        return Ok(records);
    }

    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        DangKyThi? record = await dbContext.DangKyThi
            .Include(item => item.KyThi)
            .Include(item => item.SinhVien)
            .FirstOrDefaultAsync(item => item.DangKyThiID == id);

        return record is null ? NotFound(Error("Registration not found.")) : Ok(record);
    }

    [HttpPost]
    public async Task<IActionResult> Create([FromBody] JsonElement body)
    {
        DangKyThi? record = ReadBody<DangKyThi>(body, "dang_ky_thi");
        if (record is null)
        {
            return UnprocessableEntity(Errors("Invalid registration payload."));
        }

        KyThi? kyThi = await dbContext.KyThi.FindAsync(record.KyThiID);
        SinhVien? sinhVien = await dbContext.SinhVien.FindAsync(record.SinhVienID);
        if (kyThi is null || sinhVien is null)
        {
            return NotFound(Error("Exam or student not found."));
        }

        if (kyThi.TrangThai != "published")
        {
            return UnprocessableEntity(Error("Only published exams can accept registrations."));
        }

        record.TrangThaiDangKy = string.IsNullOrWhiteSpace(record.TrangThaiDangKy) ? "registered" : record.TrangThaiDangKy;
        record.SoBaoDanh = string.IsNullOrWhiteSpace(record.SoBaoDanh) ? await GenerateSoBaoDanh(record.KyThiID) : record.SoBaoDanh;

        dbContext.DangKyThi.Add(record);

        try
        {
            await dbContext.SaveChangesAsync();
            return CreatedAtAction(nameof(Show), new { id = record.DangKyThiID }, record);
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("This student is already registered for this exam or the exam number already exists."));
        }
    }

    [HttpPatch("{id:int}")]
    [HttpPut("{id:int}")]
    public async Task<IActionResult> Update(int id, [FromBody] JsonElement body)
    {
        DangKyThi? record = await dbContext.DangKyThi.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error("Registration not found."));
        }

        DangKyThi? payload = ReadBody<DangKyThi>(body, "dang_ky_thi");
        if (payload is null)
        {
            return UnprocessableEntity(Errors("Invalid registration payload."));
        }

        record.KyThiID = payload.KyThiID;
        record.SinhVienID = payload.SinhVienID;
        record.SoBaoDanh = payload.SoBaoDanh;
        record.TrangThaiDangKy = payload.TrangThaiDangKy;
        record.CapNhatLuc = DateTime.Now;

        try
        {
            await dbContext.SaveChangesAsync();
            return Ok(record);
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Duplicate registration or exam number."));
        }
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Destroy(int id)
    {
        DangKyThi? record = await dbContext.DangKyThi.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error("Registration not found."));
        }

        dbContext.DangKyThi.Remove(record);
        try
        {
            await dbContext.SaveChangesAsync();
            return Ok(new { message = "Deleted successfully." });
        }
        catch (DbUpdateException)
        {
            return Conflict(Error("Cannot delete because this registration is already used in room assignment, seating, or attendance."));
        }
    }

    [HttpPatch("{id:int}/cancel")]
    public async Task<IActionResult> Cancel(int id)
    {
        DangKyThi? record = await dbContext.DangKyThi.Include(item => item.KyThi).FirstOrDefaultAsync(item => item.DangKyThiID == id);
        if (record is null)
        {
            return NotFound(Error("Registration not found."));
        }

        if (record.KyThi?.TrangThai != "published")
        {
            return UnprocessableEntity(Error("Registration can only be cancelled before room assignment."));
        }

        record.TrangThaiDangKy = "cancelled";
        record.CapNhatLuc = DateTime.Now;
        await dbContext.SaveChangesAsync();
        return Ok(record);
    }

    private async Task<string> GenerateSoBaoDanh(int kyThiID)
    {
        int nextNumber = await dbContext.DangKyThi.CountAsync(record => record.KyThiID == kyThiID) + 1;
        return $"SBD{kyThiID.ToString().PadLeft(3, '0')}{nextNumber.ToString().PadLeft(3, '0')}";
    }
}
